legistar/events.py

```diff
@@ -11,6 +12,49 @@
 
 
 class LegistarEventsScraper(LegistarScraper):
+    ECOMMENT_JS_URL = 'https://metro.granicusideas.com/meetings.js'
+
+    @property
+    def ecomment_dict(self):
+        """
+        Parse event IDs and eComment links from JavaScript file with lines like:
+        activateEcomment('750', '138A085F-0AC1-4A33-B2F3-AC3D6D9F710B', 'https://metro.granicusideas.com/meetings/750-finance-budget-and-audit-committee-on-2020-03-16-5-00-pm-test');
+        """
+        if not getattr(self, '_ecomment_dict', None):
+            ecomment_dict = {}
+
+            response = self.get(self.ECOMMENT_JS_URL)
+            tree = esprima.parseScript(response.text)
+
+            try:
```

Comment:
Both the exception handling and traversal of the parsed script feel pretty clunky! Any suggestions for improvement are welcome. Here's what the parsed script looks like:

```python
{
    type: "Program",
    sourceType: "script",
    body: [
        {
            type: "FunctionDeclaration",
            expression: False,
            isAsync: False,
            id: {
                type: "Identifier",
                name: "activateEcomment"
            },
            params: [
                {
                    type: "Identifier",
                    name: "event_id"
                },
                {
                    type: "Identifier",
                    name: "event_uid"
                },
                {
                    type: "Identifier",
                    name: "url"
                }
            ],
            body: {
                type: "BlockStatement",
                body: [
                    {
                        type: "VariableDeclaration",
                        declarations: [
                            {
                                type: "VariableDeclarator",
                                id: {
                                    type: "Identifier",
                                    name: "ele"
                                },
                                init: {
                                    type: "CallExpression",
                                    callee: {
                                        type: "Identifier",
                                        name: "gQuery"
                                    },
                                    arguments: [
                                        {
                                            type: "BinaryExpression",
                                            operator: "+",
                                            left: {
                                                type: "BinaryExpression",
                                                operator: "+",
                                                left: {
                                                    type: "BinaryExpression",
                                                    operator: "+",
                                                    left: {
                                                        type: "BinaryExpression",
                                                        operator: "+",
                                                        left: {
                                                            type: "BinaryExpression",
                                                            operator: "+",
                                                            left: {
                                                                type: "BinaryExpression",
                                                                operator: "+",
                                                                left: {
                                                                    type: "Literal",
                                                                    value: "[rel='ecomment'][data-event-id='",
                                                                    raw: "\"[rel='ecomment'][data-event-id='\""
                                                                },
                                                                right: {
                                                                    type: "Identifier",
                                                                    name: "event_id"
                                                                }
                                                            },
                                                            right: {
                                                                type: "Literal",
                                                                value: "'],[rel='ecomment'][data-event-id='",
                                                                raw: "\"'],[rel='ecomment'][data-event-id='\""
                                                            }
                                                        },
                                                        right: {
                                                            type: "CallExpression",
                                                            callee: {
                                                                type: "MemberExpression",
                                                                computed: False,
                                                                object: {
                                                                    type: "Identifier",
                                                                    name: "event_uid"
                                                                },
                                                                property: {
                                                                    type: "Identifier",
                                                                    name: "toUpperCase"
                                                                }
                                                            },
                                                            arguments: []
                                                        }
                                                    },
                                                    right: {
                                                        type: "Literal",
                                                        value: "'],[rel='ecomment'][data-event-id='",
                                                        raw: "\"'],[rel='ecomment'][data-event-id='\""
                                                    }
                                                },
                                                right: {
                                                    type: "CallExpression",
                                                    callee: {
                                                        type: "MemberExpression",
                                                        computed: False,
                                                        object: {
                                                            type: "Identifier",
                                                            name: "event_uid"
                                                        },
                                                        property: {
                                                            type: "Identifier",
                                                            name: "toLowerCase"
                                                        }
                                                    },
                                                    arguments: []
                                                }
                                            },
                                            right: {
                                                type: "Literal",
                                                value: "']",
                                                raw: "\"']\""
                                            }
                                        }
                                    ]
                                }
                            }
                        ],
                        kind: "var"
                    },
                    {
                        type: "ExpressionStatement",
                        expression: {
                            type: "CallExpression",
                            callee: {
                                type: "MemberExpression",
                                computed: False,
                                object: {
                                    type: "CallExpression",
                                    callee: {
                                        type: "MemberExpression",
                                        computed: False,
                                        object: {
                                            type: "Identifier",
                                            name: "ele"
                                        },
                                        property: {
                                            type: "Identifier",
                                            name: "attr"
                                        }
                                    },
                                    arguments: [
                                        {
                                            type: "Literal",
                                            value: "href",
                                            raw: "\"href\""
                                        },
                                        {
                                            type: "Identifier",
                                            name: "url"
                                        }
                                    ]
                                },
                                property: {
                                    type: "Identifier",
                                    name: "show"
                                }
                            },
                            arguments: []
                        }
                    },
                    {
                        type: "VariableDeclaration",
                        declarations: [
                            {
                                type: "VariableDeclarator",
                                id: {
                                    type: "Identifier",
                                    name: "insiteLink"
                                },
                                init: {
                                    type: "CallExpression",
                                    callee: {
                                        type: "Identifier",
                                        name: "gQuery"
                                    },
                                    arguments: [
                                        {
                                            type: "BinaryExpression",
                                            operator: "+",
                                            left: {
                                                type: "BinaryExpression",
                                                operator: "+",
                                                left: {
                                                    type: "BinaryExpression",
                                                    operator: "+",
                                                    left: {
                                                        type: "BinaryExpression",
                                                        operator: "+",
                                                        left: {
                                                            type: "BinaryExpression",
                                                            operator: "+",
                                                            left: {
                                                                type: "BinaryExpression",
                                                                operator: "+",
                                                                left: {
                                                                    type: "Literal",
                                                                    value: "form[name='aspnetForm'] .ecomment[data-event-id='",
                                                                    raw: "\"form[name='aspnetForm'] .ecomment[data-event-id='\""
                                                                },
                                                                right: {
                                                                    type: "Identifier",
                                                                    name: "event_id"
                                                                }
                                                            },
                                                            right: {
                                                                type: "Literal",
                                                                value: "'],form[name='aspnetForm'] .ecomment[data-event-id='",
                                                                raw: "\"'],form[name='aspnetForm'] .ecomment[data-event-id='\""
                                                            }
                                                        },
                                                        right: {
                                                            type: "CallExpression",
                                                            callee: {
                                                                type: "MemberExpression",
                                                                computed: False,
                                                                object: {
                                                                    type: "Identifier",
                                                                    name: "event_uid"
                                                                },
                                                                property: {
                                                                    type: "Identifier",
                                                                    name: "toLowerCase"
                                                                }
                                                            },
                                                            arguments: []
                                                        }
                                                    },
                                                    right: {
                                                        type: "Literal",
                                                        value: "'],form[name='aspnetForm'] .ecomment[data-event-id='",
                                                        raw: "\"'],form[name='aspnetForm'] .ecomment[data-event-id='\""
                                                    }
                                                },
                                                right: {
                                                    type: "CallExpression",
                                                    callee: {
                                                        type: "MemberExpression",
                                                        computed: False,
                                                        object: {
                                                            type: "Identifier",
                                                            name: "event_uid"
                                                        },
                                                        property: {
                                                            type: "Identifier",
                                                            name: "toUpperCase"
                                                        }
                                                    },
                                                    arguments: []
                                                }
                                            },
                                            right: {
                                                type: "Literal",
                                                value: "']",
                                                raw: "\"']\""
                                            }
                                        }
                                    ]
                                }
                            }
                        ],
                        kind: "var"
                    },
                    {
                        type: "ExpressionStatement",
                        expression: {
                            type: "CallExpression",
                            callee: {
                                type: "MemberExpression",
                                computed: False,
                                object: {
                                    type: "CallExpression",
                                    callee: {
                                        type: "MemberExpression",
                                        computed: False,
                                        object: {
                                            type: "CallExpression",
                                            callee: {
                                                type: "MemberExpression",
                                                computed: False,
                                                object: {
                                                    type: "CallExpression",
                                                    callee: {
                                                        type: "MemberExpression",
                                                        computed: False,
                                                        object: {
                                                            type: "CallExpression",
                                                            callee: {
                                                                type: "MemberExpression",
                                                                computed: False,
                                                                object: {
                                                                    type: "CallExpression",
                                                                    callee: {
                                                                        type: "MemberExpression",
                                                                        computed: False,
                                                                        object: {
                                                                            type: "CallExpression",
                                                                            callee: {
                                                                                type: "MemberExpression",
                                                                                computed: False,
                                                                                object: {
                                                                                    type: "Identifier",
                                                                                    name: "insiteLink"
                                                                                },
                                                                                property: {
                                                                                    type: "Identifier",
                                                                                    name: "text"
                                                                                }
                                                                            },
                                                                            arguments: [
                                                                                {
                                                                                    type: "Literal",
                                                                                    value: "eComment",
                                                                                    raw: "\"eComment\""
                                                                                }
                                                                            ]
                                                                        },
                                                                        property: {
                                                                            type: "Identifier",
                                                                            name: "attr"
                                                                        }
                                                                    },
                                                                    arguments: [
                                                                        {
                                                                            type: "Literal",
                                                                            value: "href",
                                                                            raw: "\"href\""
                                                                        },
                                                                        {
                                                                            type: "Identifier",
                                                                            name: "url"
                                                                        }
                                                                    ]
                                                                },
                                                                property: {
                                                                    type: "Identifier",
                                                                    name: "attr"
                                                                }
                                                            },
                                                            arguments: [
                                                                {
                                                                    type: "Literal",
                                                                    value: "target",
                                                                    raw: "\"target\""
                                                                },
                                                                {
                                                                    type: "Literal",
                                                                    value: "_blank",
                                                                    raw: "\"_blank\""
                                                                }
                                                            ]
                                                        },
                                                        property: {
                                                            type: "Identifier",
                                                            name: "css"
                                                        }
                                                    },
                                                    arguments: [
                                                        {
                                                            type: "Literal",
                                                            value: "color",
                                                            raw: "\"color\""
                                                        },
                                                        {
                                                            type: "Literal",
                                                            value: "blue",
                                                            raw: "\"blue\""
                                                        }
                                                    ]
                                                },
                                                property: {
                                                    type: "Identifier",
                                                    name: "css"
                                                }
                                            },
                                            arguments: [
                                                {
                                                    type: "Literal",
                                                    value: "text-decoration",
                                                    raw: "\"text-decoration\""
                                                },
                                                {
                                                    type: "Literal",
                                                    value: "underline",
                                                    raw: "\"underline\""
                                                }
                                            ]
                                        },
                                        property: {
                                            type: "Identifier",
                                            name: "css"
                                        }
                                    },
                                    arguments: [
                                        {
                                            type: "Literal",
                                            value: "cursor",
                                            raw: "\"cursor\""
                                        },
                                        {
                                            type: "Literal",
                                            value: "pointer",
                                            raw: "\"pointer\""
                                        }
                                    ]
                                },
                                property: {
                                    type: "Identifier",
                                    name: "addClass"
                                }
                            },
                            arguments: [
                                {
                                    type: "Literal",
                                    value: "active",
                                    raw: "\"active\""
                                }
                            ]
                        }
                    }
                ]
            },
            generator: False
        },
        {
            type: "ExpressionStatement",
            expression: {
                type: "CallExpression",
                callee: {
                    type: "Identifier",
                    name: "gQuery"
                },
                arguments: [
                    {
                        type: "FunctionExpression",
                        expression: False,
                        isAsync: False,
                        params: [],
                        body: {
                            type: "BlockStatement",
                            body: [
                                {
                                    type: "ExpressionStatement",
                                    expression: {
                                        type: "CallExpression",
                                        callee: {
                                            type: "Identifier",
                                            name: "activateEcomment"
                                        },
                                        arguments: [
                                            {
                                                type: "Literal",
                                                value: "752",
                                                raw: "'752'"
                                            },
                                            {
                                                type: "Literal",
                                                value: "8ADA023E-1316-4675-8479-607EC2FC8E73",
                                                raw: "'8ADA023E-1316-4675-8479-607EC2FC8E73'"
                                            },
                                            {
                                                type: "Literal",
                                                value: "https://metro.granicusideas.com/meetings/752-finance-budget-and-audit-committee-on-2020-03-17-11-00-am-test",
                                                raw: "'https://metro.granicusideas.com/meetings/752-finance-budget-and-audit-committee-on-2020-03-17-11-00-am-test'"
                                            }
                                        ]
                                    }
                                },
                                {
                                    type: "ExpressionStatement",
                                    expression: {
                                        type: "CallExpression",
                                        callee: {
                                            type: "Identifier",
                                            name: "activateEcomment"
                                        },
                                        arguments: [
                                            {
                                                type: "Literal",
                                                value: "751",
                                                raw: "'751'"
                                            },
                                            {
                                                type: "Literal",
                                                value: "E6DC5C58-D682-4B53-B78D-35E3B70AC7CA",
                                                raw: "'E6DC5C58-D682-4B53-B78D-35E3B70AC7CA'"
                                            },
                                            {
                                                type: "Literal",
                                                value: "https://metro.granicusideas.com/meetings/751-finance-budget-and-audit-committee-on-2020-03-17-12-00-pm-test",
                                                raw: "'https://metro.granicusideas.com/meetings/751-finance-budget-and-audit-committee-on-2020-03-17-12-00-pm-test'"
                                            }
                                        ]
                                    }
                                }
                            ]
                        },
                        generator: False
                    }
                ]
            }
        }
    ]
}
```

Reply:
hi @hancush, good job!

Here's a cleaner way of doing this.

```
import esprima

js = """function activateEcomment(event_id, event_uid, url) {
  // For MediaManager
  var ele = gQuery("[rel='ecomment'][data-event-id='" + event_id + "'],[rel='ecomment'][data-event-id='" + event_uid.toUpperCase() + "'],[rel='ecomment'][data-event-id='" + event_uid.toLowerCase() + "']");
  ele.attr("href", url).show();

  // For Insite
  var insiteLink = gQuery("form[name='aspnetForm'] .ecomment[data-event-id='" + event_id + "'],form[name='aspnetForm'] .ecomment[data-event-id='" + event_uid.toLowerCase() + "'],form[name='aspnetForm'] .ecomment[data-event-id='" + event_uid.toUpperCase() + "']");

  insiteLink
    .text("eComment")
    .attr("href", url)
    .attr("target", "_blank")
    .css("color", "blue")
    .css("text-decoration", "underline")
    .css("cursor", "pointer")
    .addClass("active");
}

  gQuery(function () {
    activateEcomment('751', 'E6DC5C58-D682-4B53-B78D-35E3B70AC7CA', 'https://metro.granicusideas.com/meetings/751-finance-budget-and-audit-committee-on-2020-03-17-12-00-pm-test');
  });
"""


def activateEcomment_args(js_text):
    results = []

    def is_activateComment(node, metadata):
        if node.callee and node.callee.name == 'activateEcomment':
            results.append([arg.value for arg in node.arguments])

    esprima.parse(js_text, delegate=is_activateComment)

    return results

print(activateEcomment_args(js))
```

Reply:
here's a docs for what I did here. https://esprima.readthedocs.io/en/latest/syntactic-analysis.html#example-console-calls-removal